integration/restart_test.go

```diff
@@ -139,12 +140,20 @@ func TestContainerdRestart(t *testing.T) {
 			waitCh, err := task.Wait(ctx)
 			require.NoError(t, err)
 
-			// NOTE: CRI-plugin setups watcher for each container and
-			// cleanups container when the watcher returns exit event.
-			// We just need to kill that sandbox and wait for exit
-			// event from waitCh. If the sandbox container exits,
-			// the state of sandbox must be NOT_READY.
-			require.NoError(t, task.Kill(ctx, syscall.SIGKILL, containerd.WithKillAll))
+			err = task.Kill(ctx, syscall.SIGKILL, containerd.WithKillAll)
+			if goruntime.GOOS != "windows" {
+				// NOTE: CRI-plugin setups watcher for each container and
+				// cleanups container when the watcher returns exit event.
+				// We just need to kill that sandbox and wait for exit
+				// event from waitCh. If the sandbox container exits,
+				// the state of sandbox must be NOT_READY.
+				require.NoError(t, err)
+			} else {
+				// NOTE(gsamfira): There is no task running on Windows. See above NOTE from
+				// claudiub. On windows, there are no running containers in the
+				// "notready-sandbox" array, so this will always return ErrorNotFound.
+				assert.True(t, errdefs.IsNotFound(err), err)
```

Comment:
I don't understand the comment. The NOT-READY state means that we should stop the sandbox container and we can get NOT-Ready state after restart containerd. Even if there is no running containers, we are handling the sandbox one. It is confusing

Reply:
You are right. This comment was confusing. Sorry about that.

Sending a ```Kill()``` to a task that is already stopped, or that was never started, will return an ```ErrorNotFound``` on Windows. This sandbox only has a container in stopped state and one in created state. So sending a kill here will always return an ```ErrorNotFound``` on Windows. By the time we send this ```Kill()```, the wait channel for the task should already be closed and an ```ErrorNotFound``` can be safely ignored.

On Linux, there is also a container in started state, added here: https://github.com/containerd/containerd/blob/main/integration/restart_test.go#L83-L91

so on Linux, the ```Kill()``` function should always return a ```nil``` error.

Sorry for the confusion. I rewrote the comment. Hopefully it's clear now.

Reply:
On Windows, is the sandbox container running after RunPodSandbox? Even if there is no running containers.

Reply:
After ```containerd``` restart, we end up with one sandbox in ```SANDBOX_READY``` and another one in ```SANDBOX_NOTREADY```. This is checked here:

https://github.com/containerd/containerd/blob/main/integration/restart_test.go#L175-L181

so the test case should be fulfilled. Unless I'm reading it wrong 😀.

Another option is to simply not try and send a kill there, as there is no running container, if you prefer.

Reply:
> After containerd restart, we end up with one sandbox in SANDBOX_READY and another one in SANDBOX_NOTREADY.

Before restart, how to change the running sandbox container into exited state? We should send signal, right? If so, why we can receive not-found error?

Reply:
Let me backtrack a bit and attempt to clarify.

The test assumes that killing the sandbox, will not kill the containers running inside it. This is not true on Windows. On Windows, killing the sandbox will kill all tasks inside the sandbox, and by extension any containers. This is the reason why a container in ```RUNNING``` state is not added, unless ```runtime.GOOS != "windows"``` here:

https://github.com/containerd/containerd/blob/main/integration/restart_test.go#L83-L85

If we add a container that is expected to be in running state after killing the sandbox and restarting ```containerd```, then that will fail on Windows.

@claudiubelu Do you think this test makes sense on Windows, based on the assumptions described here?

Reply:
> The test assumes that killing the sandbox, will not kill the containers running inside it. This is not true on Windows. On Windows, killing the sandbox will kill all tasks inside the sandbox, and by extension any containers.

I understand that. But let's checkout the change. The L155 in your change is to check the error is not-found. The sandbox container is running so you have to send signal to let the sandbox exit. So that not-found is expected error when you kill running windows sandbox container?

Reply:
If the sandbox has no running task inside it, yes. On windows, if it's a sandbox, it gets the pod, then forwards the ```Kill``` signal to any task inside the pod. There is no running task, so it returns not-found. If you have at least one running task, it will return nil. But of course, the test will fail because after ```containerd``` restart, the sandbox is expected to have one running container. Which is not the case on Windows.

Pinged Claudiu Belu to weigh in on whether this test makes sense on Windows given it's current assumptions. If it's not, then the change I made will be removed or rewritten to conform with expectations. Marking the PR as a WiP.

Reply:
> If you have at least one running task, it will return nil. But of course, the test will fail because after containerd restart, the sandbox is expected to have one running container. Which is not the case on Windows.

Thanks for your patience! Good to know this! 🙏 thanks! TIL things😂

Reply:
My pleasure! And thanks for the review! I get worried when nobody questions the changes I make :laughing:

Reply:
I did a small test. It seems that after killing the sandbox on linux, it's not transitioned to ```NOTREADY```. Listing sandboxes after the kill and printing the status shows that both are ready. After containerd restart, the second one is ```NOTREADY```, which is to be expected. On Windows, the seciond sanbox is marked as NOTREADY due to the fact that no running container is inside it. So this test should be relevant by simply ignoring the not-found error on windows.

Removing the WiP. This should fix the failure in the CI, as soon as the following 2 PRs in hcsshim are merged:

https://github.com/microsoft/hcsshim/pull/1249
https://github.com/microsoft/hcsshim/pull/1252

and the hcsshim dependency in containerd is updated.

Reply:
> Listing sandboxes after the kill and printing the status shows that both are ready.

Did you wait for a while? The state is updated in async way.

https://github.com/containerd/containerd/blob/653f2f1ce4b8eaee13810fd0c5fce73e039d14a6/pkg/cri/server/events.go#L418

Reply:
I added logging to the test that showed the status was unchanged by the time containerd restarted. Should we change the test and poll until the sandbox transitions to ```NOTREADY``` before restarting containerd?

Reply:
No need to do that. The `task.Wait` return can guarantee that. :)

Reply:
This is the change I made:

https://paste.samfira.com/public/p/TNTzRQIaJREB14JFwJ7i7NIS

This is the test output:

https://paste.samfira.com/public/p/mXhv2fZLvmgxJjapkKd2LG4h

Reply:
Yeah. It is about timing. :joy: